### ISSUE TITLE
Consider adding support for formatting strings with ANSI escape codes

### ISSUE THREAD

**Issue by DavidGregory084** [typelevel/paiges#134]
I have spent a little bit of time playing around with a combination of this library and lihaoyi/fansi, however I have noticed that the use of ANSI coloured strings affects the formatting of documents, e.g.

Without coloured text:
```
module Test.Typechecker.Apply { let lam = a -> a; let app = lam(true) }
```
With coloured text
```
module Test.Typechecker.Apply {
  let lam = a -> a;
  let app = lam(
      true
    )
}
```
It looks great though!

It might be quite hard to figure out what the non-coloured length of the String is so this might not be worth the effort, but I thought I'd ask whether you guys think there is a good way to approach this?

**Comment by DavidGregory084** [comment]
One option would be to regex out all the escape codes before calculating the length of a String, i.e. `s.replaceAll("\u001b\\[[;\\d]*m", "")`, but I'm certain this would have a very upsetting effect on the performance of the library

**Comment by olafurpg** [comment]
I think #91 is related, you could attach colors to the docs and insert the ANSI escapes after rendering. I agree this would be nice functionality, I have similar use-cases such as rendering syntax highlighted code in html with hyperlinks to symbol definitions.

**Comment by johnynek** [comment]
I guess we need to understand that some characters are 0-width in the rendering.

I guess this could be done by having a new `Doc` type that wraps 0 width characters. I think it should probably be doable. Happy to help coach anyone through a PR.

**Comment by johnynek** [comment]
closed by #184
